File: docs/deriver.js
// Mirrors src/deriver.py. Turns the raw parsed report into the data dict the
// snapshot template consumes (with all derived calculations).

import { lookupBonusRates } from './bonus.js';

const COLORS = ['#0F6E56', '#185FA5', '#D85A30', '#534AB7', '#BA7517', '#0C447C', '#993556'];

const ASSET_CLASS_LABELS = {
  'Global Equity': 'Global equity',
  'Asia Equity': 'Asia equity',
  'US Equity': 'US equity',
  'Emerging Markets Bond': 'EM bond',
  'Multi Asset': 'Multi-asset',
  'Asia Pacific Equity': 'Asia-Pacific equity',
  'European Equity': 'European equity',
  'Japan Equity': 'Japan equity',
  'Global Bond': 'Global bond',
  'Asia Bond': 'Asia bond',
};


/**
 * Derive the full data model for the snapshot template.
 *
 * @param {object} raw         Output of parsePdf()
 * @param {object} [options]   Bonus-exclusion adjustments
 * @param {boolean} [options.excludeWelcomeBonus=false]
 * @param {boolean} [options.excludeAnnualPremiumBonus=false]
 * @param {number}  [options.welcomeBonusRate]      Override auto-detected welcome rate (0-1)
 * @param {number}  [options.annualPremiumBonusRate] Override auto-detected annual rate (0-1)
 *
 * When either exclude flag is true, Net gain, Total return, Annualised IRR,
 * and the Capital/Gains bar are recomputed by adding the bonus dollar amounts
 * to the cost basis (treating bonuses as if they were premiums paid).
 */
export function derive(raw, options = {}) {
  const flexiTerm = intFrom(raw.policyName, /Flexi\s+(\d+)/) ?? 10;
  const policyTermYears = intFrom(raw.policyName, /(\d+)\s+Years?\s+Flexi/) ?? flexiTerm;

  const issueDate = parseDmy(raw.policyIssueDate);
  const reportDate = parseFlexibleDate(raw.reportDate);

  const monthsInvested = monthsBetween(issueDate, reportDate);
  const premiumsPaidCount = countAnniversariesPaid(issueDate, reportDate);
  const inferred = inferAnnualPremium(
    raw.policyInvestmentCost,
    issueDate,
    reportDate,
    premiumsPaidCount,
  );
  const annualPremium = inferred.annualPremium;
  const premiumFrequency = inferred.frequency;             // 'monthly' | 'annual'
  const premiumFrequencyAmbiguous = inferred.ambiguous;     // true at 12/24/36-mo boundaries
  const premiumsRemaining = Math.max(0, flexiTerm - premiumsPaidCount);

  // Auto-detect bonus rates from the product/variation; allow caller overrides.
  // For unrecognised products (e.g. Manulink Investor (II) - SRS) we treat the
  // bonus rates as zero and IGNORE any caller overrides — the panel rates are
  // configured for the *recognised* policy and shouldn't bleed onto Manulink
  // policies in the same PDF, where there is no equivalent first-year bonus.
  const auto = lookupBonusRates(raw.product, raw.variation, annualPremium);
  const welcomeBonusRate = auto.recognised
    ? (options.welcomeBonusRate ?? auto.welcomeRate)
    : 0;
  let annualPremiumBonusRate = auto.recognised
    ? (options.annualPremiumBonusRate ?? auto.annualPremiumRate)
    : 0;

  // Manulife rule: the InvestReady (III) annual premium bonus is paid only
  // when premiums are collected ANNUALLY. Monthly-pay clients forfeit it
  // entirely (the welcome bonus still applies). InvestReady Growth's 3 %
  // annual bonus is unconditional — it stays even on monthly-pay because
  // there's no first-year clawback. Only zero it out when we're CONFIDENT
  // the policy is monthly; in the ambiguous (12/24/36-month boundary) case
  // we leave the bonus in and surface a disclaimer to the user instead.
  if (raw.product === 'InvestReady (III)'
      && premiumFrequency === 'monthly'
      && !premiumFrequencyAmbiguous) {
    annualPremiumBonusRate = 0;
  }

  const welcomeBonusAmount = round2(annualPremium * welcomeBonusRate);
  const annualPremiumBonusAmount = round2(annualPremium * annualPremiumBonusRate);

  const excludeWelcome = !!options.excludeWelcomeBonus;
  const excludeAnnual = !!options.excludeAnnualPremiumBonus;
  const excludedBonusAmount =
    (excludeWelcome ? welcomeBonusAmount : 0) +
    (excludeAnnual ? annualPremiumBonusAmount : 0);
  const adjustmentsActive = excludedBonusAmount > 0;

  // Adjusted cost = real cost + bonuses we treat as cost. When no exclusions,
  // adjustedCost === policyInvestmentCost and the metric values fall back to
  // the PDF's stated numbers (preserves what Manulife officially reported).
  const adjustedCost = raw.policyInvestmentCost + excludedBonusAmount;

  let totalPnlDollar = raw.totalPnlDollar;
  let totalPnlPct = raw.totalPnlPct;
  let annualisedPnlPct = raw.annualisedPnlPct;
  if (adjustmentsActive && raw.accountValue && adjustedCost > 0) {
    totalPnlDollar = round2(raw.accountValue - adjustedCost);
    totalPnlPct = round2((raw.accountValue / adjustedCost - 1) * 100);
    const days = Math.max(1, daysBetween(issueDate, reportDate));
    annualisedPnlPct = round2(
      (Math.pow(raw.accountValue / adjustedCost, 365 / days) - 1) * 100
    );
  }

  // capitalPct can exceed 100 when account value is below cost (a loss).
  // Clamp the visual representation to [0, 100] so the progress bar fills
  // sensibly, and surface a separate `lossPct` for the legend label.
  const rawCapitalPct = raw.accountValue
    ? (adjustedCost / raw.accountValue) * 100
    : 0;
  const inLoss = rawCapitalPct > 100;
  const capitalPct = Math.min(100, rawCapitalPct);
  const gainsPct = Math.max(0, 100 - capitalPct);
  const lossPct = inLoss ? rawCapitalPct - 100 : 0;

  const allEnriched = enrichHoldings(raw.holdings, raw.accountValue);
  // Manulife reports per-fund Total P&L as a CUMULATIVE figure that includes
  // realised gains from switch-outs. So a fund the customer largely switched
  // away from can show a tiny current Fund Value sitting next to a P&L
  // figure many multiples larger — accurate per Manulife's accounting, but
  // confusing in a one-page summary aimed at clients.
  //
  // Heuristic: a holding is "leftover" from a fund switch when both
  //   - it's currently <0.5% of the account value, AND
  //   - its absolute P&L is at least as large as its current value.
  // We split such positions out into `minorHoldings`; the snapshot template
  // hides them from the donut and the per-fund table and surfaces a compact
  // footnote line so the visible rows tell a coherent story.
  const isLeftoverSwitch = (h) =>
    h.allocationPct < 0.5 && Math.abs(h.pnlDollar) > h.fundValue;
  const holdingsEnriched = allEnriched.filter(h => !isLeftoverSwitch(h));
  const minorHoldings = allEnriched.filter(isLeftoverSwitch);

  // Re-color the visible holdings in their (now contiguous) order so the
  // colours stay paired with the visual ranking on screen.
  holdingsEnriched.forEach((h, i) => { h.color = COLORS[i % COLORS.length]; });

  const minorHoldingsValue = minorHoldings.reduce((s, h) => s + h.fundValue, 0);
  const minorHoldingsPnl = minorHoldings.reduce((s, h) => s + h.pnlDollar, 0);

  // fundPnlTotal still includes minor positions — it must reconcile with
  // Manulife's "Grand Total" P&L line, which sums every fund regardless of
  // current size.
  const fundPnlTotal = allEnriched.reduce((s, h) => s + h.pnlDollar, 0);

  // Estimate the market gain on the bonus units themselves.
  // Bonuses were credited as units in the funds at inception. Their growth
  // since then ≈ (basic-premium return rate) × bonus principal. This makes
  // the gap between policy-level Net gain and the per-fund P&L total
  // explicit (the fund table only sums P&L on basic-premium allocations).
  const totalBonusPrincipal = welcomeBonusAmount + annualPremiumBonusAmount;
  const basicPremiumReturnRate = raw.policyInvestmentCost > 0
    ? fundPnlTotal / raw.policyInvestmentCost
    : 0;
  const bonusGrowthEstimate = round2(totalBonusPrincipal * basicPremiumReturnRate);

  const equityPct = holdingsEnriched
    .filter(h => /equity/i.test(h.assetClass) || /equity/i.test(h.subAssetClass))
    .reduce((s, h) => s + h.allocationPct, 0);
  const incomePct = Math.max(0, 100 - equityPct);

  return {
    customerName: raw.customerName,
    customerNameTitle: titleCaseName(raw.customerName),
    reportDate: raw.reportDate,
    policyName: raw.policyName,
    policyNamePretty: prettifyPolicyName(raw.policyName),
    policyNumber: raw.policyNumber,
    policyIssueDate: raw.policyIssueDate,
    inceptionDatePretty: prettyDate(raw.policyIssueDate),
    accountValue: raw.accountValue,
    policyInvestmentCost: raw.policyInvestmentCost,
    totalPnlDollar,
    totalPnlPct,
    annualisedPnlPct,
    totalRiderPremiums: raw.totalRiderPremiums,
    totalDividendsReinvested: raw.totalDividendsReinvested,
    riskProfile: raw.riskProfile,
    ckaStatus: raw.ckaStatus,
    ckaExpiry: raw.ckaExpiry,
    ckaExpiryPretty: prettyDate(raw.ckaExpiry),
    flexiTerm,
    policyTermYears,
    monthsInvested,
    premiumsPaidCount,
    annualPremium,
    premiumsRemaining,
    capitalPct: round1(capitalPct),
    gainsPct: round1(gainsPct),
    inLoss,
    lossPct: round1(lossPct),
    holdingsEnriched,
    minorHoldings,
    minorHoldingsValue,
    minorHoldingsPnl,
    fundPnlTotal,
    equityPct: round1(equityPct),
    incomePct: round1(incomePct),
    // Bonus context (consumed by the template footnote)
    product: raw.product,
    variation: raw.variation,
    welcomeBonusRate,
    annualPremiumBonusRate,
    welcomeBonusAmount,
    annualPremiumBonusAmount,
    totalBonusPrincipal,
    bonusGrowthEstimate,
    excludeWelcomeBonus: excludeWelcome,
    excludeAnnualPremiumBonus: excludeAnnual,
    adjustmentsActive,
    // Frequency context (consumed by the snapshot for the disclaimer line)
    premiumFrequency,
    premiumFrequencyAmbiguous,
  };
}


// ---------- Calculation helpers ----------------------------------------------

function countAnniversariesPaid(issueDate, reportDate) {
  if (reportDate < issueDate) return 0;
  let years = reportDate.getFullYear() - issueDate.getFullYear();
  const nextAnniv = new Date(issueDate);
  nextAnniv.setFullYear(issueDate.getFullYear() + years);
  if (reportDate >= nextAnniv) return years + 1;
  return years;
}

// The PDF doesn't state the premium payment frequency (monthly / quarterly /
// annual), so we have to infer it from the math. Two interpretations are
// possible for any (totalPaid, monthsElapsed, anniversaries) triple:
//
//   ANNUAL  — premium collected once per anniversary year.
//             count = anniversaries (incl. inception)
//             rate  = totalPaid / count          ← already the annual premium
//
//   MONTHLY — premium collected at inception then every month after.
//             count = monthsElapsed + 1          (inception is a payment date)
//             rate  = totalPaid / count          ← monthly rate
//             annualised = rate × 12
//
// Real-world Manulife premiums are multiples of $100 (often $500 / $1,000),
// so pick whichever interpretation lands closer to a clean $100 multiple.
// A bug we hit in the wild: a monthly $833.33 / $10,000-a-year policy was
// mis-inferred as 2 annual installments of $9,583.32, dropping the welcome
// bonus from the upper tier (45 %) to the lower tier (15 %) and roughly
// halving the displayed first-year bonus.
//
// Bias toward annual on near-ties — annual is the more common case and the
// PDF's structure (premium paid by anniversary) leans that way too. The
// monthly inference only wins when it's BOTH very clean (≤ $5 from a $100
// multiple) AND meaningfully cleaner than the annual one (by > $5).
// Use completed monthly payment intervals here; the display month count
// rounds partial months up, which is too generous for payment counts.
//
// Genuine ambiguity exists when totalPaid happens to equal `annual × N` for
// both interpretations — i.e. when monthlyPayments is itself a multiple of
// 12 (months elapsed = 11, 23, 35, ...). At those boundaries an annual-pay
// $10K and a monthly-pay $833.33 client both produce the same totalPaid.
// We can't disambiguate, so we return frequency='annual' (the safer default
// for the bonus calc) and set ambiguous=true so the snapshot can warn.
//
// Returns: { annualPremium, frequency: 'monthly'|'annual', ambiguous: bool }
function inferAnnualPremium(totalPaid, issueDate, reportDate, anniversaries) {
  if (totalPaid <= 0) {
    return { annualPremium: 0, frequency: 'annual', ambiguous: false };
  }

  const annualEst = anniversaries > 0 ? totalPaid / anniversaries : 0;

  const monthsElapsed = fullMonthsBetween(issueDate, reportDate);
  const monthlyPayments = monthsElapsed + 1;
  const monthlyAnnualEst = monthlyPayments > 0
    ? (totalPaid / monthlyPayments) * 12
    : 0;

  const annualResidual = residualFromHundred(annualEst);
  const monthlyResidual = residualFromHundred(monthlyAnnualEst);

  // Monthly inference is the clear winner.
  if (monthlyAnnualEst > 0
      && monthlyResidual <= 5
      && monthlyResidual < annualResidual - 5) {
    // Round to the nearest $100 to absorb sub-cent drift in the per-payment
    // math (e.g. $833.331 × 12 = $9,999.97 → $10,000).
    return {
      annualPremium: Math.round(monthlyAnnualEst / 100) * 100,
      frequency: 'monthly',
      ambiguous: false,
    };
  }

  // Both interpretations land on the SAME clean number (within a few dollars).
  // This is the 12/24/36-month-boundary case — math can't tell the two apart.
  const ambiguous = (
    annualEst > 0 && monthlyAnnualEst > 0
    && annualResidual <= 5 && monthlyResidual <= 5
    && Math.abs(annualEst - monthlyAnnualEst) <= 5
  );
  // When the annual estimate is itself within $5 of a clean $100 multiple,
  // snap to that multiple — absorbs sub-cent drift like $9,999.96 → $10,000.
  // Otherwise preserve the exact value (e.g. an unusual $10,250 stays put).
  const cleanAnnual = annualResidual <= 5
    ? Math.round(annualEst / 100) * 100
    : round2(annualEst);

  return {
    annualPremium: cleanAnnual,
    frequency: 'annual',
    ambiguous,
  };
}

function residualFromHundred(v) {
  return Math.abs(v - Math.round(v / 100) * 100);
}

function fullMonthsBetween(start, end) {
  if (end < start) return 0;
  let months = (end.getFullYear() - start.getFullYear()) * 12 + (end.getMonth() - start.getMonth());
  if (end.getDate() < start.getDate()) months -= 1;
  return Math.max(0, months);
}

function daysBetween(start, end) {
  const ms = end.getTime() - start.getTime();
  return Math.floor(ms / (24 * 60 * 60 * 1000));
}

function monthsBetween(start, end) {
  if (end < start) return 0;
  let months = (end.getFullYear() - start.getFullYear()) * 12 + (end.getMonth() - start.getMonth());
  if (end.getDate() < start.getDate()) months -= 1;
  // Round up when more than half the trailing month has passed (matches Python deriver).
  const trailingDays = (end.getDate() - start.getDate() + 31) % 31;
  if (trailingDays >= 15) months += 1;
  return Math.max(0, months);
}


// ---------- Holdings enrichment ----------------------------------------------

function enrichHoldings(holdings, accountValue) {
  const sorted = [...holdings].sort((a, b) => b.fundValue - a.fundValue);
  return sorted.map((h, i) => ({
    ...h,
    allocationPct: accountValue ? (h.fundValue / accountValue) * 100 : 0,
    color: COLORS[i % COLORS.length],
    displayName: shortenFundName(h.fundFullName),
    assetClassLabel: ASSET_CLASS_LABELS[h.subAssetClass] ?? h.subAssetClass,
  }));
}

// Display the fund name EXACTLY as Manulife wrote it in the PDF — share
// class, currency, hedge markers, distribution codes and all. The only
// thing this function does is patch up PDF.js word-level extraction
// artefacts so compound words don't render with a stray space:
//   "Bh- SGD"      → "Bh-SGD"
//   "Multi- Asset" → "Multi-Asset"
//   "H2- SGD"      → "H2-SGD"
// We deliberately preserve real separator dashes ("Manulife Global Fund - …")
// because they're part of the canonical name. Earlier versions of this
// function tried to "shorten" the name by stripping share-class tails;
// that lost meaningful descriptors (Hard Currency, Diversified Income, Opps)
// and ended up trimming things the user wanted preserved. The user's
// reasonable expectation: type the full name exactly as the report shows it.
export function shortenFundName(full) {
  let s = (full || '').trim();
  // Re-fuse only when the dash/slash is glued to the preceding token
  // (PDF.js artefact signature: no space before, space after). Real
  // separator dashes — " - " — have spaces on both sides and stay intact.
  s = s.replace(/(\S[\/\-])\s+(\S)/g, '$1$2');
  return s.trim();
}


// ---------- Date helpers ------------------------------------------------------

function parseDmy(s) {
  // "01/04/2024" -> Date
  const [d, m, y] = s.split('/').map(Number);
  return new Date(y, m - 1, d);
}

function parseFlexibleDate(s) {
  // "30 Apr 2026" or "01/04/2024" or "2026-04-30"
  const trimmed = s.trim();
  if (/^\d{2}\/\d{2}\/\d{4}$/.test(trimmed)) return parseDmy(trimmed);
  const d = new Date(trimmed);
  if (!isNaN(d)) return d;
  throw new Error(`Unparseable date: ${s}`);
}

function prettyDate(s) {
  if (!s) return '';
  try {
    const d = parseDmy(s);
    const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
    return `${String(d.getDate()).padStart(2, '0')} ${months[d.getMonth()]} ${d.getFullYear()}`;
  } catch {
    return s;
  }
}


// ---------- Misc helpers ------------------------------------------------------

function intFrom(text, re) {
  const m = re.exec(text || '');
  return m ? parseInt(m[1], 10) : null;
}

function titleCaseName(name) {
  return (name || '').trim().split(/\s+/).map(w => w.charAt(0).toUpperCase() + w.slice(1).toLowerCase()).join(' ');
}

function prettifyPolicyName(name) {
  return (name || '').replace(/\(III\)\s+/, '(III) — ');
}

function round1(n) { return Math.round(n * 10) / 10; }
function round2(n) { return Math.round(n * 100) / 100; }

File: src/deriver.py
"""Turn a RawReport into the full data dict the Jinja2 template expects."""

from __future__ import annotations

import re
from dataclasses import asdict
from datetime import date
from typing import Any

from dateutil.parser import parse as parse_date
from dateutil.relativedelta import relativedelta

from .parser import Holding, RawReport


COLORS = ["#0F6E56", "#185FA5", "#D85A30", "#534AB7", "#BA7517", "#0C447C", "#993556"]

SUB_ASSET_CLASS_LABELS = {
    "Global Equity": "Global equity",
    "Asia Equity": "Asia equity",
    "US Equity": "US equity",
    "Emerging Markets Bond": "EM bond",
    "Multi Asset": "Multi-asset",
    "Asia Pacific Equity": "Asia-Pacific equity",
    "European Equity": "European equity",
    "Japan Equity": "Japan equity",
    "Global Bond": "Global bond",
    "Asia Bond": "Asia bond",
}

EQUITY_KEYWORDS = ("equity",)


def derive(raw: RawReport) -> dict[str, Any]:
    is_single_premium = _is_single_premium_policy(raw.policy_name)
    flexi_term = 0 if is_single_premium else (_int_from(raw.policy_name, r"Flexi\s+(\d+)") or 10)
    policy_term_years = None if is_single_premium else (_int_from(raw.policy_name, r"(\d+)\s+Years?\s+Flexi") or flexi_term)

    issue_date = _parse_dmy(raw.policy_issue_date)
    report_date = parse_date(raw.report_date).date()

    months_invested = _months_between(issue_date, report_date)
    premiums_paid_count = 1 if is_single_premium else count_anniversaries_paid(issue_date, report_date)
    inferred = (
        {"annual_premium": raw.policy_investment_cost, "frequency": "single", "ambiguous": False}
        if is_single_premium
        else infer_annual_premium(
            raw.policy_investment_cost,
            issue_date,
            report_date,
            premiums_paid_count,
        )
    )
    annual_premium = inferred["annual_premium"]
    premium_frequency = inferred["frequency"]                # 'monthly' | 'annual'
    premium_frequency_ambiguous = inferred["ambiguous"]
    premiums_remaining = 0 if is_single_premium else max(0, flexi_term - premiums_paid_count)

    # capital_pct can exceed 100 when account < cost (a loss).
    # Clamp the visual representation to [0, 100] so the progress bar fills
    # sensibly, and surface a separate `loss_pct` for the legend label.
    raw_capital_pct = (
        (raw.policy_investment_cost / raw.account_value * 100) if raw.account_value else 0.0
    )
    in_loss = raw_capital_pct > 100
    capital_pct = min(100.0, raw_capital_pct)
    gains_pct = max(0.0, 100 - capital_pct)
    loss_pct = max(0.0, raw_capital_pct - 100) if in_loss else 0.0

    holdings_enriched = _enrich_holdings(raw.holdings, raw.account_value)
    fund_pnl_total = sum(h["pnl_dollar"] for h in holdings_enriched)

    equity_pct = sum(
        h["allocation_pct"] for h in holdings_enriched
        if any(kw in h["asset_class"].lower() or kw in h["sub_asset_class"].lower() for kw in EQUITY_KEYWORDS)
    )
    income_pct = max(0.0, 100 - equity_pct)

    return {
        "customer_name": raw.customer_name,
        "customer_name_title": _title_case_name(raw.customer_name),
        "report_date": raw.report_date,
        "policy_name": raw.policy_name,
        "policy_name_pretty": _prettify_policy_name(raw.policy_name),
        "policy_number": raw.policy_number,
        "policy_issue_date": raw.policy_issue_date,
        "inception_date_pretty": _pretty_date(raw.policy_issue_date),
        "account_value": raw.account_value,
        "policy_investment_cost": raw.policy_investment_cost,
        "total_pnl_dollar": raw.total_pnl_dollar,
        "total_pnl_pct": raw.total_pnl_pct,
        "annualised_pnl_pct": raw.annualised_pnl_pct,
        "total_rider_premiums": raw.total_rider_premiums,
        "total_dividends_reinvested": raw.total_dividends_reinvested,
        "total_dividends_paid_out": raw.total_dividends_paid_out,
        "total_partial_withdrawal": raw.total_partial_withdrawal,
        "risk_profile": raw.risk_profile,
        "risk_profile_expiry": raw.risk_profile_expiry,
        "risk_profile_expiry_pretty": _pretty_date(raw.risk_profile_expiry),
        "risk_profile_expired": _is_expired_dmy(raw.risk_profile_expiry, report_date),
        "cka_status": raw.cka_status,
        "cka_expiry": raw.cka_expiry,
        "cka_expiry_pretty": _pretty_date(raw.cka_expiry),
        "cka_expired": _is_expired_dmy(raw.cka_expiry, report_date),
        "is_single_premium": is_single_premium,
        "flexi_term": flexi_term,
        "policy_term_years": policy_term_years,
        "months_invested": months_invested,
        "premiums_paid_count": premiums_paid_count,
        "annual_premium": annual_premium,
        "premium_frequency": premium_frequency,
        "premium_frequency_ambiguous": premium_frequency_ambiguous,
        "premiums_remaining": premiums_remaining,
        "capital_pct": round(capital_pct, 1),
        "gains_pct": round(gains_pct, 1),
        "in_loss": in_loss,
        "loss_pct": round(loss_pct, 1),
        "holdings_enriched": holdings_enriched,
        "fund_pnl_total": fund_pnl_total,
        "equity_pct": round(equity_pct, 1),
        "income_pct": round(income_pct, 1),
    }


def count_anniversaries_paid(issue_date: date, report_date: date) -> int:
    """Inception itself counts as the first premium."""
    if report_date < issue_date:
        return 0
    years_diff = relativedelta(report_date, issue_date).years
    next_anniv = issue_date + relativedelta(years=years_diff)
    if report_date >= next_anniv:
        return years_diff + 1
    return years_diff


def infer_annual_premium(
    total_paid: float, issue_date: date, report_date: date, anniversaries: int
) -> dict[str, Any]:
    """Infer the policy's annual premium and the inferred payment frequency.

    The PDF doesn't state the premium frequency, so we back it out by trying
    both interpretations:

      ANNUAL  - one premium per anniversary year. Annual = total / anniversaries.
      MONTHLY - one premium at inception then every month after.
                Monthly rate = total / (months_elapsed + 1); annual = rate * 12.

    Real-world Manulife premiums are multiples of $100 (often $500 / $1,000),
    so we pick whichever interpretation lands closer to a clean $100 multiple.
    Monthly only wins when its residual is BOTH <= $5 AND > $5 better than
    annual's. Bias toward annual on near-ties (more common case).
    Use completed monthly payment intervals here; the display month count
    rounds partial months up, which is too generous for payment counts.

    Genuine ambiguity exists at 12 / 24 / 36-month boundaries — at those
    moments an annual-pay $10K and a monthly-pay $833.33 client both produce
    the same total_paid. The math can't tell them apart. We return the
    annual interpretation (safer for the bonus calc) and set
    ambiguous=True so the snapshot can render a disclaimer.

    Returns: {'annual_premium': float, 'frequency': 'monthly'|'annual',
              'ambiguous': bool}
    """
    if total_paid <= 0:
        return {"annual_premium": 0.0, "frequency": "annual", "ambiguous": False}

    annual_est = total_paid / anniversaries if anniversaries > 0 else 0.0

    months_elapsed = _full_months_between(issue_date, report_date)
    monthly_payments = months_elapsed + 1
    monthly_annual_est = (
        (total_paid / monthly_payments) * 12 if monthly_payments > 0 else 0.0
    )

    annual_residual = _residual_from_hundred(annual_est)
    monthly_residual = _residual_from_hundred(monthly_annual_est)

    if (
        monthly_annual_est > 0
        and monthly_residual <= 5
        and monthly_residual < annual_residual - 5
    ):
        return {
            "annual_premium": float(round(monthly_annual_est / 100) * 100),
            "frequency": "monthly",
            "ambiguous": False,
        }

    ambiguous = (
        annual_est > 0
        and monthly_annual_est > 0
        and annual_residual <= 5
        and monthly_residual <= 5
        and abs(annual_est - monthly_annual_est) <= 5
    )
    # Snap to the nearest $100 when within $5 — absorbs sub-cent drift like
    # $9,999.96 → $10,000. Preserve unusual values (e.g. $10,250) verbatim.
    clean_annual = (
        float(round(annual_est / 100) * 100) if annual_residual <= 5
        else round(annual_est, 2)
    )

    return {
        "annual_premium": clean_annual,
        "frequency": "annual",
        "ambiguous": ambiguous,
    }


def _residual_from_hundred(v: float) -> float:
    return abs(v - round(v / 100) * 100)


def _full_months_between(start: date, end: date) -> int:
    """Completed monthly payment intervals between two dates."""
    if end < start:
        return 0
    rd = relativedelta(end, start)
    return max(0, rd.years * 12 + rd.months)


def _int_from(text: str, pattern: str) -> int | None:
    m = re.search(pattern, text or "")
    return int(m.group(1)) if m else None


def _parse_dmy(s: str) -> date:
    return parse_date(s, dayfirst=True).date()


def _months_between(start: date, end: date) -> int:
    """Calendar months elapsed, rounded up when more than half the month has passed."""
    if end < start:
        return 0
    rd = relativedelta(end, start)
    months = rd.years * 12 + rd.months
    if rd.days >= 15:
        months += 1
    return months


def _title_case_name(name: str) -> str:
    return " ".join(w.capitalize() for w in (name or "").strip().split())


def _pretty_date(s: str) -> str:
    """'18/08/2026' -> '18 Aug 2026'."""
    if not s:
        return ""
    try:
        d = _parse_dmy(s)
    except (ValueError, TypeError):
        return s
    return d.strftime("%d %b %Y")


def _prettify_policy_name(name: str) -> str:
    """'Manulife InvestReady (III) 13 Years Flexi 10' -> 'Manulife InvestReady (III) — 13 Years Flexi 10'."""
    return re.sub(r"\(III\)\s+", "(III) — ", name or "")


def _is_single_premium_policy(policy_name: str) -> bool:
    return bool(re.search(r"Manulink\s+Investor\s*\(II\)\s*-\s*SRS", policy_name or "", re.IGNORECASE))


def _is_expired_dmy(s: str, report_date: date) -> bool:
    if not s:
        return False
    try:
        return _parse_dmy(s) < report_date
    except (ValueError, TypeError):
        return False


def _enrich_holdings(holdings: list[Holding], account_value: float) -> list[dict]:
    items = sorted(
        (asdict(h) for h in holdings),
        key=lambda h: h["fund_value"],
        reverse=True,
    )
    for i, h in enumerate(items):
        h["allocation_pct"] = (h["fund_value"] / account_value * 100) if account_value else 0.0
        h["color"] = COLORS[i % len(COLORS)]
        h["display_name"] = shorten_fund_name(h["fund_full_name"])
        h["sub_asset_class_label"] = SUB_ASSET_CLASS_LABELS.get(h["sub_asset_class"], h["sub_asset_class"])
    return items


def shorten_fund_name(full: str) -> str:
    """Display the fund name EXACTLY as Manulife wrote it.

    This function used to "shorten" the verbose name by stripping
    share-class / hedge / distribution suffixes — but every iteration of
    that lost meaningful descriptors (Hard Currency, Diversified Income,
    Opps, …) and the user reasonably wants the canonical name preserved.
    The only thing we do here is patch up PDF.js word-level extraction
    artefacts so compound words don't render with a stray space:

        'Bh- SGD'      -> 'Bh-SGD'
        'Multi- Asset' -> 'Multi-Asset'
        'H2- SGD'      -> 'H2-SGD'

    Real separator dashes (' - ') are preserved verbatim because
    'Manulife Global Fund - Global Multi-Asset Diversified Income Fund
    AA (SGD Hedged) MDIST (G)' is the canonical name as printed in the
    PDF and that's exactly what the report should display.
    """
    s = full.strip()
    # Re-fuse only when the dash/slash is glued to the preceding token —
    # PDF.js artefact signature is no-space-before / space-after. Real
    # separator dashes (' - ') stay intact because they have spaces on
    # both sides.
    s = re.sub(r"(\S[\/\-])\s+(\S)", r"\1\2", s)
    return s.strip()
